src/goto-instrument/contracts/dynamic-frames/dfcc_loop_nesting_graph.cpp

```diff
@@ -0,0 +1,92 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking
+
+Author: Qinheping Hu, qinhh@amazon.com
+
+Author: Remi Delmas, delmasrd@amazon.com
+
+Date: March 2023
+
+\*******************************************************************/
+
+#include "dfcc_loop_nesting_graph.h"
+
+#include <util/invariant.h>
+#include <util/message.h>
+
+dfcc_loop_nesting_graph_nodet::dfcc_loop_nesting_graph_nodet(
+  const goto_programt::targett head,
+  const goto_programt::targett latch,
+  const loopt instructions)
+  : head(head), latch(latch), instructions(instructions)
+{
+}
+
+/// \pre The function \ref dfcc_check_loop_normal_form must have successfully
+/// been applied to the model.
+dfcc_loop_nesting_grapht
+build_loop_nesting_graph(goto_programt &goto_program, messaget &log)
+{
+  natural_loops_mutablet natural_loops(goto_program);
+  dfcc_loop_nesting_grapht loop_nesting_graph;
+  // Add a node per natural loop
+  for(auto &loop : natural_loops.loop_map)
+  {
+    auto loop_head = loop.first;
+    auto &loop_instructions = loop.second;
+
+    if(loop_instructions.size() <= 1)
+    {
+      // ignore single instruction loops of the form
+      // LOOP: goto LOOP;
+      continue;
+    }
+
+    // Find the instruction that jumps back to `loop_head` and has the
+    // highest GOTO location number, define it as the latch.
+    auto loop_latch = loop_head;
+    for(const auto &t : loop_instructions)
+    {
+      if(
+        t->is_goto() && t->get_target() == loop_head &&
+        t->location_number > loop_latch->location_number)
+        loop_latch = t;
+    }
+
+    if(loop_latch == loop_head)
+    {
+      log.error() << "Could not find latch instruction of loop starting at: "
+                  << loop_head->source_location() << messaget::eom;
+      throw analysis_exceptiont("Could not find loop latch instruction");
+    }
```

Comment:
I'd expect a `DATA_INVARIANT` here, unless this is a scenario that's actually possible in GOTO programs.

Reply:
Agreed and fixed.